libs/libviewer/src/dataset.rs

```diff
@@ -0,0 +1,295 @@
+
+        for file in &self.files {
+            let (metadata, num_rows) = match file.num_rows {
+                Some(rows) => (None, rows),
+                None => {
+                    let metadata =
+                        read_metadata(self.metadata_store.clone(), file.metadata_path.as_ref())
+                            .await?;
+                    let num_rows = metadata.file_metadata().num_rows() as u64;
+                    (Some(metadata), num_rows)
+                }
+            };
+
+            // here we handle file pruning based on the offset and limit
+            if current_offset + num_rows > offset {
+                // this file must be scanned, calculate the offset within the file
+                let file_offset = if offset > current_offset {
+                    offset - current_offset
+                } else {
+                    0
+                };
+
+                // calculate the limit for this file
+                let file_limit = remaining_limit.min(num_rows - file_offset);
+                remaining_limit -= file_limit;
+
+                let metadata = match metadata {
+                    Some(meta) => meta,
+                    None => {
+                        read_metadata(self.metadata_store.clone(), file.metadata_path.as_ref())
+                            .await?
+                    }
+                };
+                scans.push(ParquetScan {
+                    file: file.clone(),
+                    limit: file_limit,
+                    offset: file_offset,
+                    metadata: metadata,
+                });
+
+                if remaining_limit == 0 {
+                    break;
+                }
+            }
+
+            current_offset += num_rows;
+        }
+        Ok(scans)
+    }
+
+    pub async fn index(&self, files: Option<&[IndexedFile]>) -> Result<()> {
+        let tasks = files
+            .unwrap_or(&self.files)
+            .iter()
+            .map(async move |file| {
+                let metadata = read_metadata(self.data_store.clone(), file.path.as_ref()).await?;
+                write_metadata(
+                    metadata,
+                    self.metadata_store.clone(),
+                    file.metadata_path.as_ref(),
+                )
+                .await
+            })
+            .collect::<Vec<_>>();
+
+        future::try_join_all(tasks).await?;
+
+        Ok(())
+    }
+
+    pub async fn scan(
+        &self,
+        limit: Option<u64>,
+        offset: Option<u64>,
+    ) -> Result<(Vec<RecordBatch>, Vec<IndexedFile>)> {
+        // 1. create an object reader for each file in the access plan
+        // 2. generate a stream of record batches from each reader
+        // 3. flatten the streams into a single stream
+        // 4. collect the record batches into a single vector
+
+        let plan = self.plan(limit, offset).await?;
+        let files_to_index: Vec<IndexedFile> = plan
```

Comment:
maybe add a comment to say that it's not used ? (or remove it)

Reply:
Removed it.